manager/allocator/cnmallocator/portallocator.go

```diff
@@ -117,16 +118,40 @@ func newPortAllocator() (*portAllocator, error) {
 	return &portAllocator{portSpaces: portSpaces}, nil
 }
 
+type (
+	// these are deliberately aliases, otherwise we'd only match the same type, not signature.
+	legacyConstructor = func(ds datastore.DataStore, id string, start, end uint64) (*idm.Idm, error)
+	idmConstructor    = func(id string, start, end uint64) (*idm.Idm, error)
```

Comment:
🤦 that took me some time to understand "why it doesn't match".... got bitten by Go's strong typing;

This will only match a function that's an actual "one", not "any function with signature `func(string)`

```go
type one func(string)
```

See below (notice that both `one` and `three` are allowed in the switch .. because, well, they're not the same:
code below will match "two", not "one": https://go.dev/play/p/L_J_To1xRhp

```go
package main

import "fmt"

type one func(string)
type two = func(string)
type three func(string)

func myFunc(string) {}

func assertType(fn any) {
	switch f := fn.(type) {
	case one:
		fmt.Printf("one: %T == %T", f, fn)
	case two:
		fmt.Printf("two: %T == %T", f, fn)
	case three:
		fmt.Printf("three: %T == %T", f, fn)
	default:
		fmt.Printf("unknown: %T != %T", f, fn)
	}
}

func main() {
	assertType(myFunc)
}
```